ambry-store/src/main/java/com/github/ambry/store/CompactionLog.java

```diff
@@ -213,6 +309,111 @@ void setSafeToken(StoreFindToken safeToken) {
         cycleLog.compactionDetails);
   }
 
+  /**
+   * Add a pair of index segment start offsets. The first offset has to be an index segment start offset that belongs to
+   * an under compaction log segment. The second offset hast to be an index segment start offset that belongs to a log
+   * segment created this compaction.
+   * @param before The index segment start offset before compaction
+   * @param after The index segment start offset after compaction
+   */
+  void addIndexSegmentOffsetPair(Offset before, Offset after) {
+    if (!isIndexSegmentOffsetsPersisted()) {
+      return;
+    }
+    // If the before offset already exists, it means the index segment has already been copied, or half way through
+    // copying. Either way, we should keep the original after offset, since the after offset should reference to the
+    // first index segment that has data from before index segment.
+    if (beforeAndAfterIndexSegmentOffsets.containsKey(before)) {
+      logger.trace("{}: Offset {} already exist in the map", file, before);
+      return;
+    }
+    LogSegmentName logSegmentName = before.getName();
+    if (!getCompactionDetails().getLogSegmentsUnderCompaction().contains(logSegmentName)) {
+      throw new IllegalArgumentException("Offset is not part of the under compaction log segments: " + before);
+    }
+    // TODO: Can we find a way to validate the after offset?
+    beforeAndAfterIndexSegmentOffsets.put(before, after);
+    logger.trace("{}: Add offsets pair to {}: {}", file, before, after);
+    flush();
+  }
+
+  /**
+   * Update all the index segment offset values added through method {@link #addIndexSegmentOffsetPair}. If the input is
+   * null, the index segment offsets map would be cleared.
+   * @param validOffset The valid {@link Offset} to replace all the previous invalid {@link Offset}.
+   */
+  void updateIndexSegmentOffsetsWithValidOffset(Offset validOffset) {
+    if (!isIndexSegmentOffsetsPersisted()) {
+      return;
+    }
+    logger.info("{}: Validate offsets map with: {}", file, validOffset);
+    if (validOffset == null) {
+      beforeAndAfterIndexSegmentOffsets.clear();
+    } else {
+      for (Offset before : beforeAndAfterIndexSegmentOffsets.keySet()) {
+        beforeAndAfterIndexSegmentOffsets.put(before, validOffset);
+      }
+    }
+    flush();
+  }
+
+  /**
+   * Return the index segment offset map. Each key and value is before and after pair passed to {@link #addIndexSegmentOffsetPair}.
+   * DON'T modify this map.
```

Comment:
nit: Would `Collections.unmodifiableMap(map)` work?

Reply:
problem is that unmodifiableMap returns a Map object, what I need here is a NavigableMap.

Reply:
Can we use Collections.unmodifiableNavigableMap?

Reply:
@SophieGuo410 good catch, didn't know that exists.